### PR TITLE
[SYCL] Align GEMM dispatch

### PR DESCRIPTION
**Pull request by airMeng**
The issue is exposed during https://github.com/ggerganov/llama.cpp/pull/6408

I will split #6408 into several smaller PRs for eazy reviewing, the tasks will be updated according to issues exposed

- [x] GEMM aligning with CUDA backends
- [ ] remove all global variables and pass the context instead
- [ ] leave the async copy to common backend
- [ ] separate GEMM files outside of ggml-sycl.c for maintainability.

### CODE REVIEW

File: ggml-sycl.cpp [hunk 1]
```diff
@@ -15189,75 +15208,59 @@ static void ggml_sycl_mul_mat(const ggml_tensor * src0, const ggml_tensor * src1
...
+        && src1->type == GGML_TYPE_F32 && dst->type == GGML_TYPE_F32;
+
+    // fp16 performance always better on gen12+
+    const bool fp16_performance_good = true;
+
+    // mmvq and mmq need the __dp4a instruction which is available for gen12+
+    use_mul_mat_vec_q = use_mul_mat_vec_q; // Check dp4a
+    use_mul_mat_q     = use_mul_mat_q    ; // check dp4a
+#ifdef SYCL_USE_XMX
+    use_mul_mat_q     = use_mul_mat_q     && (!fp16_performance_good || src1->ne[1] <= MMQ_MAX_BATCH_SIZE);
```
Comment (joeatodd) [inline ggml-sycl.cpp:15240]: The logic in this block is a little confusing:
`(!fp16_performance_good || src1->ne[1] <= MMQ_MAX_BATCH_SIZE)`.

It says: use MMQ if either 1) FP16 perf is bad, or 2) the number of columns in `src1` is less than or equal to the maximum.

Is this the intention?

**Review by joeatodd — COMMENTED** [review]
This is a welcome addition since it simplifies the MatMul dispatch :smile:

We spotted a couple of issues with the code, I've added comments. I hope they're helpful.

File: ggml-sycl.cpp [hunk 2]
```diff
@@ -15173,6 +15173,25 @@ catch (sycl::exception const &exc) {
   std::exit(1);
 }
 
+bool ggml_sycl_supports_mmq(enum ggml_type type) {
+    // TODO: accuracy issues in MMQ
+    return false;
```
Comment (AidanBeltonS) [inline ggml-sycl.cpp:15177]: Could you elaborate on what accuracy issues you are having with MMQ?

File: CMakeLists.txt [hunk 3]
```diff
@@ -96,8 +96,8 @@ option(LLAMA_LLAMAFILE                       "llama: use llamafile SGEMM"
 set(LLAMA_BLAS_VENDOR "Generic" CACHE STRING "llama: BLAS library vendor")
 option(LLAMA_CUDA                            "llama: use CUDA"                                  OFF)
 option(LLAMA_CUBLAS                          "llama: use CUDA (deprecated, use LLAMA_CUDA)"     OFF)
-option(LLAMA_CUDA_FORCE_DMMV                 "llama: use dmmv instead of mmvq CUDA kernels"     OFF)
-option(LLAMA_CUDA_FORCE_MMQ                  "llama: use mmq kernels instead of cuBLAS"         OFF)
+option(LLAMA_FORCE_DMMV                      "llama: use dmmv instead of mmvq CUDA kernels"     OFF)
+option(LLAMA_FORCE_MMQ                       "llama: use mmq kernels instead of cuBLAS"         OFF)
```
[the remark this answers is not in the document]
Comment (airMeng) [reply CMakeLists.txt]: nice catch

File: ggml-sycl.cpp [hunk 4]
```diff
@@ -15189,75 +15208,59 @@ static void ggml_sycl_mul_mat(const ggml_tensor * src0, const ggml_tensor * src1
         }
     }
 
+#if !defined(GGML_SYCL_FORCE_MMQ)
+    #define SYCL_USE_XMX
```
[the remark this answers is not in the document]
Comment (airMeng) [reply ggml-sycl.cpp]: just align with CUDA logic

File: CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (joeatodd) [reply CMakeLists.txt]: Also these settings are described in the README, so should probably be updated there.

File: ggml-sycl.cpp (hunk 1 above)
Comment (airMeng) [reply ggml-sycl.cpp:15229]: just align with CUDA logic

**Review by AidanBeltonS — COMMENTED** [review]
I am seeing a new test failures on `Arc A770`, is this to be expected?

```
  MUL_MAT(type_a=iq2_xxs,type_b=f32,m=16,n=1,k=256,bs=[1,1],nr=[1,1]): GGML_ASSERT: /builds/perseus-performance-libraries/llama_ci/llama.cpp/ggml-sycl.cpp:13858: false
ggml_sycl_op_dequantize_mul_mat_vec unsupported GGML_TYPE 16
```

**Comment by airMeng** [comment]
> I am seeing a new test failures on `Arc A770`, is this to be expected?
> 
> ```
>   MUL_MAT(type_a=iq2_xxs,type_b=f32,m=16,n=1,k=256,bs=[1,1],nr=[1,1]): GGML_ASSERT: /builds/perseus-performance-libraries/llama_ci/llama.cpp/ggml-sycl.cpp:13858: false
> ggml_sycl_op_dequantize_mul_mat_vec unsupported GGML_TYPE 16
> ```

fixed

File: ggml-sycl.cpp [hunk 5]
```diff
@@ -15209,76 +15246,42 @@ static void ggml_sycl_mul_mat(const ggml_tensor * src0, const ggml_tensor * src1
...
-    const bool use_xmx = true;
-#else
-    const bool use_xmx = false;
-#endif
+    // check data types and tensor shapes for custom matrix multiplication kernels:
+    bool use_dequantize_mul_mat_vec = ggml_sycl_supports_dmmv(src0->type)
+        && src1->type == GGML_TYPE_F32 && dst->type == GGML_TYPE_F32
+        && src0->ne[0] % GGML_SYCL_DMMV_X == 0 && src1->ne[1] == 1;
+
+    bool          use_mul_mat_vec_q =  ggml_is_quantized(src0->type)
```
Comment (NeoZhangJianyu) [inline ggml-sycl.cpp:15254]:
```suggestion
    bool use_mul_mat_vec_q =  ggml_is_quantized(src0->type)
```

File: ggml-sycl.cpp [hunk 6]
```diff
@@ -15209,76 +15246,42 @@ static void ggml_sycl_mul_mat(const ggml_tensor * src0, const ggml_tensor * src1
...
+        && src1->ne[1] <= MMVQ_MAX_BATCH_SIZE;
 
-    // debug helpers
-    //printf("src0: %8d %8d %8d %8d\n", src0->ne[0], src0->ne[1], src0->ne[2], src0->ne[3]);
-    //printf("      %8d %8d %8d %8d\n", src0->nb[0], src0->nb[1], src0->nb[2], src0->nb[3]);
-    //printf("src1: %8d %8d %8d %8d\n", src1->ne[0], src1->ne[1], src1->ne[2], src1->ne[3]);
-    //printf("      %8d %8d %8d %8d\n", src1->nb[0], src1->nb[1], src1->nb[2], src1->nb[3]);
-    //printf("src0 is contiguous %d, transposed %d, type = %s, name = %s\n", ggml_is_contiguous(src0), ggml_is_transposed(src0), ggml_type_name(src0->type), src0->name);
-    //printf("src1 is contiguous %d, transposed %d, type = %s, name = %s\n", ggml_is_contiguous(src1), ggml_is_transposed(src1), ggml_type_name(src1->type), src1->name);
+    bool              use_mul_mat_q =  ggml_sycl_supports_mmq(src0->type)
```
Comment (NeoZhangJianyu) [inline ggml-sycl.cpp:15258]:
```suggestion
    bool use_mul_mat_q =  ggml_sycl_supports_mmq(src0->type)
```

File: ggml-sycl.cpp [hunk 7]
```diff
@@ -15209,76 +15246,42 @@ static void ggml_sycl_mul_mat(const ggml_tensor * src0, const ggml_tensor * src1
...
+        && src1->type == GGML_TYPE_F32 && dst->type == GGML_TYPE_F32;
 
-    if (!split && all_on_device && !use_xmx && src0->type == GGML_TYPE_F16 && ggml_is_permuted(src0) && ggml_is_permuted(src1) && src1->ne[1] == 1) {
+    // mmvq and mmq need the __dp4a instruction which is available for gen12+
+    // Workaround in https://github.com/ggerganov/llama.cpp/commit/95f84d5ce8b449a9b16009434aca800df504a02e
+    use_mul_mat_q     = use_mul_mat_q && (src0->type != GGML_TYPE_IQ2_XXS);
```
Comment (NeoZhangJianyu) [inline ggml-sycl.cpp:15263]:
```suggestion
    use_mul_mat_q = use_mul_mat_q && (src0->type != GGML_TYPE_IQ2_XXS);
```

File: ggml-sycl.cpp [hunk 8]
```diff
@@ -15209,76 +15246,42 @@ static void ggml_sycl_mul_mat(const ggml_tensor * src0, const ggml_tensor * src1
...
+#ifdef SYCL_USE_XMX
+    use_mul_mat_q     = use_mul_mat_q     && (src1->ne[1] <= MMQ_MAX_BATCH_SIZE);
```
Comment (NeoZhangJianyu) [inline ggml-sycl.cpp:15265]:
```suggestion
    use_mul_mat_q = use_mul_mat_q && (src1->ne[1] <= MMQ_MAX_BATCH_SIZE);
```

File: ggml-sycl.cpp [hunk 9]
```diff
@@ -15193,6 +15192,44 @@ catch (sycl::exception const &exc) {
...
+    //     case GGML_TYPE_Q8_0:
+    //     case GGML_TYPE_Q2_K:
+    //     case GGML_TYPE_Q3_K:
+    //     case GGML_TYPE_Q4_K:
+    //     case GGML_TYPE_Q5_K:
+    //     case GGML_TYPE_Q6_K:
+    //         return true;
+    //     default:
+    //         return false;
+    // }
```
Comment (NeoZhangJianyu) [inline ggml-sycl.cpp:15212]: rm above unused code.

**Review by NeoZhangJianyu — APPROVED** [review]
The performance is increased from 34 token/s to 37 on Arc770 with llama2-7b-Q4.
It's good!

**Pull request merged**